cmd/commandfuncs.go

```diff
@@ -538,6 +538,47 @@ func cmdValidateConfig(fl Flags) (int, error) {
 	return caddy.ExitCodeSuccess, nil
 }
 
+func cmdFormatConfig(fl Flags) (int, error) {
+	// Default path of file is Caddyfile
+	formatCmdConfigFile := fl.Arg(0)
+	if formatCmdConfigFile == "" {
+		formatCmdConfigFile = "Caddyfile"
+	}
+
+	// Default formatter is caddyfile
+	formatCmdFormatterFlag := fl.String("formatter")
+	if formatCmdFormatterFlag == "" {
+		formatCmdFormatterFlag = "caddyfile"
+	}
+
+	formatCmdWriteFlag := fl.Bool("write")
+
+	input, err := ioutil.ReadFile(formatCmdConfigFile)
+	if err != nil {
+		return caddy.ExitCodeFailedStartup,
+			fmt.Errorf("reading input file: %v", err)
+	}
+
+	formatter := caddyconfig.GetFormatter(formatCmdFormatterFlag)
+	if formatter == nil {
+		return caddy.ExitCodeFailedStartup,
+			fmt.Errorf("unrecognized config formatter: %s", formatCmdFormatterFlag)
+	}
+
+	output := formatter.Format(input)
+
+	if formatCmdWriteFlag {
+		err = ioutil.WriteFile(formatCmdConfigFile, output, 0644)
```

Comment:
We must be careful not to change a file's permissions. Be sure the original permissions are preserved.

Reply:
This actually does not change permissions. If the file does not exist (which in this case we're sure it does) it just creates a file with these permissions.

From docs:
> WriteFile writes data to a file named by filename. If the file does not exist, WriteFile creates it with permissions perm (before umask); otherwise WriteFile truncates it before writing.

Reply:
Ah, I stand corrected. Thanks!